compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64ArrayIndexOfOp.java

```diff
@@ -77,190 +90,301 @@ public AArch64ArrayIndexOfOp(int arrayBaseOffset, JavaKind valueKind, LIRGenerat
-            // move on to next chunk if no match found and new curIndex is < 0
-            masm.adds(64, curIndex, curIndex, chunkSize);
-            masm.branchConditionally(ConditionFlag.MI, chunkedReadLoop);
-
-            /*
-             * 4. Reset the ref chars count to zero to read the last chunk, only if the chunk was
-             * not already read (i.e. the curIndex is not the chunk size). Because curIndex is not
-             * aligned to the chunk size, it likely that some indexes searched again; however, this
-             * is not a correctness problem and will not result in more than one iteration of
-             * chunkedReadLoop being executed.
-             */
-            masm.compare(32, curIndex, chunkSize);
-            masm.branchConditionally(ConditionFlag.EQ, end);
-            masm.mov(curIndex, 0);
-            masm.jmp(chunkedReadLoop);
-
-            /*
-             * 5. The searchChar found in chunk so retrieve its position relative to the current
-             * index.
-             */
-            masm.bind(charInChunk);
-            masm.rev(64, curChar, curChar);
-            masm.clz(64, curChar, curChar);
-            /*
-             * 5.3 Convert number of leading zeros to number of zero bytes and add to current index.
-             */
-            masm.add(64, curIndex, curIndex, curChar, ShiftType.LSR, 3);
+        masm.bind(processTail);
+        masm.cmp(64, chunkToReadAddress, endOfString);
+        masm.branchConditionally(ConditionFlag.GE, end);
+        masm.sub(64, chunkToReadAddress, endOfString, 32);
+        /*
+         * Move back the 'endOfString' by 32-bytes because at the end of 'searchByChunkLoopTail',
+         * the 'chunkToRead' would be reset to 32-byte aligned addressed. Thus, the
+         * searchByChunkLoopHead would never using 'chunkToReadAddress' >= 'endOfString' condition.
+         */
+        masm.mov(64, endOfString, chunkToReadAddress);
+        masm.jmp(searchByChunkLoopTail);
+
+        /* 4. If the element is found in a 32-byte chunk then find its position. */
+        masm.bind(matchInChunk);
+        /*
+         * 4.1 Using the magic constant set 2 bits in the matching byte(s) that represent its
+         * position in the chunk
+         */
+        try (ScratchRegister scratchReg = masm.getScratchRegister()) {
+            Register magicConstantReg = scratchReg.getRegister();
+            masm.mov(magicConstantReg, magicConstant);
+            masm.neon.dupVG(ASIMDSize.FullReg, ElementSize.DoubleWord, tmpRegV1, magicConstantReg);
         }
+        masm.neon.andVVV(ASIMDSize.FullReg, chunkPart1RegV, chunkPart1RegV, tmpRegV1);
+        masm.neon.andVVV(ASIMDSize.FullReg, chunkPart2RegV, chunkPart2RegV, tmpRegV1);
+        /* 4.2 Convert 32-byte to 8-byte representation, 2 bits per byte. */
+        /* Reduce from 256 -> 128 bits. */
+        masm.neon.addpVVV(ASIMDSize.FullReg, ElementSize.Byte, chunkPart1RegV, chunkPart1RegV, chunkPart2RegV);
```

Comment:
Made this code not dependent on element size; because it's finding the first matching byte I think this should be okay